libazureinit/src/imds.rs

```diff
@@ -87,65 +90,110 @@ where
     }
 }
 
-pub async fn query(client: &Client) -> Result<InstanceMetadata, Error> {
-    let url = "http://169.254.169.254/metadata/instance?api-version=2021-02-01";
+// Set of StatusCodes that should be retried,
+// e.g. 400, 404, 410, 429, 500, 503.
+const RETRY_CODES: &[StatusCode] = &[
+    StatusCode::BAD_REQUEST,
+    StatusCode::NOT_FOUND,
+    StatusCode::GONE,
+    StatusCode::TOO_MANY_REQUESTS,
+    StatusCode::INTERNAL_SERVER_ERROR,
+    StatusCode::SERVICE_UNAVAILABLE,
+];
+
+static DEFAULT_IMDS_URL: &str =
+    "http://169.254.169.254/metadata/instance?api-version=2021-02-01";
+
+pub async fn query(
+    client: &Client,
+    retry_interval: Duration,
+    total_timeout: Duration,
+    inputurl: Option<&str>,
+) -> Result<InstanceMetadata, Error> {
     let mut headers = HeaderMap::new();
 
+    let url = match inputurl {
+        Some(url) => url,
+        None => DEFAULT_IMDS_URL,
+    };
+
     headers.insert("Metadata", HeaderValue::from_static("true"));
 
-    let request = client.get(url).headers(headers);
-    let response = request.send().await?;
+    let response = timeout(total_timeout, async {
```

Comment:
@anhvoms Should the retry interval exponentially increase (By factors of 2) as failures to query imds repeatedly happen? It is my understanding that this is what cloud-init does today

Reply:
No, we want fix retry interval to minimize impact to provisioning latency. cloud-init does fixed retry interval, not exponential. We have exponential backup for logging in the case of pre-provisioning VM (because some pre-provisioned VM can stay in pre-provisioning state for very long). Excessive logging isn't a concern for azure-init